packages/frameworks-express/src/index.ts

```diff
@@ -0,0 +1,202 @@
+ *
+ * ```ts
+ * export function authenticatedUser(
+ *   req: Request,
+ *   res: Response,
+ *   next: NextFunction
+ * ) {
+ *   const session = res.locals.session ?? (await getSession(req, authConfig))
+ *   if (!session?.user) {
+ *     res.redirect("/login")
+ *   } else {
+ *     next()
+ *   }
+ * }
+ * ```
+ *
+ * ### Per Route
+ * To protect a single route, simply add the middleware to the route as follows:
+ * ```ts title="app.ts"
+ * // ...
+ *
+ * // This route is protected
+ * app.get("/profile", authenticatedUser, (req, res) => {
+ *   const { session } = res.locals
+ *   res.render("profile", { user: session?.user })
+ * })
+ *
+ * // This route is not protected
+ * app.get("/", (req, res) => {
+ *   res.render("index")
+ * })
+ *
+ * app.use("/", root)
+ * ```
+ * ### Per Group of Routes
+ * To protect a group of routes, define a router and add the middleware to the router as follows:
+ *
+ * ```ts title="routes/protected.route.ts"
+ * import { Router } from "express"
+ *
+ * const router = Router()
+ *
+ * router.use(authenticatedUser) // All routes defined after this will be protected
+ *
+ * router.get("/", (req, res) => {
+ *   res.render("protected")
+ * })
+ *
+ * export default router
+ * ```
+ *
+ * Then we mount the router as follows:
+ * ```ts title="app.ts"
+ * import protected from "./routes/protected.route"
+ * app.use("/protected", protected)
+ * ```
+ *
+ * ## Notes
+ *
+ * :::info
+ * PRs to improve this documentation are welcome! See [this file](https://github.com/nextauthjs/next-auth/blob/main/packages/frameworks-express/src/lib/index.ts).
+ * :::
+ *
+ * @module index
+ */
+
+import { Auth } from "@auth/core"
+import type { AuthConfig, Session } from "@auth/core/types"
+import { Request as ExpressRequest, Response as ExpressResponse } from "express"
+import { toWebRequest, toExpressResponse } from "./lib/index.js"
+
+function ExpressAuthHandler(authConfig: AuthConfig) {
+  return async (req: ExpressRequest, res: ExpressResponse) => {
+    const request = toWebRequest(req)
+    const response = await Auth(request, authConfig)
+    await toExpressResponse(response, res)
+  }
+}
+
+export function ExpressAuth(
+  config: AuthConfig
+): ReturnType<typeof ExpressAuthHandler> {
+  const { ...authOptions } = config
+  authOptions.secret ??= process.env.AUTH_SECRET
+  authOptions.trustHost ??= !!(
+    process.env.AUTH_TRUST_HOST ??
+    process.env.VERCEL ??
+    process.env.NODE_ENV !== "production"
+  )
+
+  const handler = ExpressAuthHandler(authOptions)
+
+  return handler
+}
+
+export type GetSessionResult = Promise<Session | null>
+
+export async function getSession(
+  req: ExpressRequest,
+  options: AuthConfig
+): GetSessionResult {
+  options.secret ??= process.env.AUTH_SECRET
+  options.trustHost ??= true
+
+  const request = toWebRequest(req)
+  const url = new URL("/api/auth/session", request.url)
+
+  const response = await Auth(
+    new Request(url, { headers: request.headers }),
```

Comment:
I think this should suffice as is 🤔? The`getSession` method extracts the authentication token from the cookies in the request headers. And then if the token is present and valid then the user is authenticated regardless of HTTP method being used. Also, recall that our implementation is a mirror of [that of solid-start's](https://github.com/nextauthjs/next-auth/blob/22434ad222b076d688939470e4a80d28b65399af/packages/frameworks-solid-start/src/index.ts#L257) with tweaks to translate the request first. Thoughts?

Reply:
I understood the CSRF checks that Auth.js provides as only occurring on sign-in/out's. Check ([this](https://authjs.dev/reference/core/errors#missingcsrf)) out. I think one would have to introduce extra CSRF protection for their application's API routes as needed. It may be possible to add this as a feature to Auth.js, but I don't believe this is supported at the moment.